File: vps_relay.go
package main

import (
	"bufio"
	"bytes"
	"encoding/binary"
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	MsgRegister  = 0x01
	MsgKeepalive = 0x02
	MsgAck       = 0x03
	MsgDataFwd   = 0x10
	MsgDataReply = 0x11

	HomeTimeout = 90 * time.Second
	PeerTimeout = 120 * time.Second
)

var Magic = []byte{0xDE, 0xAD, 0xBE, 0xEF}

type Peer struct {
	ID        uint32
	Addr      *net.UDPAddr
	PackedHdr []byte
	LastSeen  time.Time
}

type Relay struct {
	mu          sync.RWMutex
	wgPort      int
	bindAddr    string
	homeEp      *net.UDPAddr
	homeTs      time.Time
	peersByID   map[uint32]*Peer
	peersByAddr map[string]*Peer
	nextPeerID  uint32
	pktFwd      uint64
	pktRet      uint64
	drops       uint64
}

func NewRelay(bindAddr string, wgPort int) *Relay {
	return &Relay{
		wgPort:      wgPort,
		bindAddr:    bindAddr,
		peersByID:   make(map[uint32]*Peer),
		peersByAddr: make(map[string]*Peer),
		nextPeerID:  1,
	}
}

func (r *Relay) homeAlive() bool {
	return r.homeEp != nil && time.Since(r.homeTs) < HomeTimeout
}

func (r *Relay) run() {
	addr := fmt.Sprintf("%s:%d", r.bindAddr, r.wgPort)
	udpAddr, err := net.ResolveUDPAddr("udp", addr)
	if err != nil {
		log.Fatalf("Invalid bind address: %v", err)
	}

	conn, err := net.ListenUDP("udp", udpAddr)
	if err != nil {
		log.Fatalf("Failed to bind UDP: %v", err)
	}
	defer conn.Close()

	// Maximize buffer sizes for high throughput (4MB)
	conn.SetReadBuffer(4 * 1024 * 1024)
	conn.SetWriteBuffer(4 * 1024 * 1024)

	log.Printf("VPS relay running on %s", addr)

	go r.cleanupLoop()

	buf := make([]byte, 65536)
	for {
		n, raddr, err := conn.ReadFromUDP(buf)
		if err != nil {
			log.Printf("Read error: %v", err)
			continue
		}
		if n < 4 {
			continue
		}

		// Discriminate packets based on Magic prefix
		if bytes.Equal(buf[:4], Magic) {
			r.handleControl(conn, buf[:n], raddr)
		} else {
			r.handleWG(conn, buf[:n], raddr)
		}
	}
}

func (r *Relay) handleControl(conn *net.UDPConn, data []byte, addr *net.UDPAddr) {
	if len(data) < 5 {
		return
	}
	msgType := data[4]

	r.mu.Lock()
	defer r.mu.Unlock()

	switch msgType {
	case MsgRegister:
		changed := r.homeEp == nil || r.homeEp.String() != addr.String()
		r.homeEp = addr
		r.homeTs = time.Now()
		if changed {
			log.Printf("Home registered: %s", addr.String())
		}
		conn.WriteToUDP(append(Magic, MsgAck), addr)

	case MsgKeepalive:
		if r.homeEp != nil && r.homeEp.String() == addr.String() {
			r.homeTs = time.Now()
			conn.WriteToUDP(append(Magic, MsgAck), addr)
		} else {
			log.Printf("Keepalive from unknown home: %s", addr.String())
		}

	case MsgDataReply:
		if r.homeEp == nil || r.homeEp.String() != addr.String() || len(data) < 9 {
			return
		}
		peerID := binary.BigEndian.Uint32(data[5:9])
		peer, exists := r.peersByID[peerID]
		if exists {
			peer.LastSeen = time.Now()
			wgBytes := data[9:]
			_, err := conn.WriteToUDP(wgBytes, peer.Addr)
			if err != nil {
				r.drops++
			} else {
				r.pktRet++
			}
		}
	}
}

func (r *Relay) handleWG(conn *net.UDPConn, data []byte, addr *net.UDPAddr) {
	r.mu.Lock()
	if !r.homeAlive() {
		r.mu.Unlock()
		return
	}

	addrStr := addr.String()
	peer, exists := r.peersByAddr[addrStr]
	if !exists {
		peer = &Peer{
			ID:   r.nextPeerID,
			Addr: addr,
		}
		r.nextPeerID++

		// Pre-pack the header: peer_id(4) + ip(4) + port(2) = 10 bytes
		hdr := make([]byte, 10)
		binary.BigEndian.PutUint32(hdr[0:4], peer.ID)
		ipBytes := addr.IP.To4()
		if ipBytes == nil {
			ipBytes = net.ParseIP("0.0.0.0").To4() // Fallback
		}
		copy(hdr[4:8], ipBytes)
		binary.BigEndian.PutUint16(hdr[8:10], uint16(addr.Port))
		peer.PackedHdr = hdr

		r.peersByID[peer.ID] = peer
		r.peersByAddr[addrStr] = peer
		log.Printf("New WG peer %s id=%d", addrStr, peer.ID)
	}
	peer.LastSeen = time.Now()
	homeEp := r.homeEp

	// Create forward packet: MAGIC(4) | MsgDataFwd(1) | HDR(10) | DATA
	fwd := make([]byte, 15+len(data))
	copy(fwd[0:4], Magic)
	fwd[4] = MsgDataFwd
	copy(fwd[5:15], peer.PackedHdr)
	copy(fwd[15:], data)
	r.mu.Unlock()

	_, err := conn.WriteToUDP(fwd, homeEp)

	r.mu.Lock()
	if err != nil {
		r.drops++
	} else {
		r.pktFwd++
	}
	r.mu.Unlock()
}

func (r *Relay) cleanupLoop() {
	ticker := time.NewTicker(30 * time.Second)
	for range ticker.C {
		r.mu.Lock()
		now := time.Now()
		for id, p := range r.peersByID {
			if now.Sub(p.LastSeen) > PeerTimeout {
				delete(r.peersByID, id)
				delete(r.peersByAddr, p.Addr.String())
				log.Printf("Dropped idle peer %s id=%d", p.Addr.String(), p.ID)
			}
		}

		homeStr := "✗ (none)"
		if r.homeAlive() {
			homeStr = fmt.Sprintf("✓ %s", r.homeEp.String())
		}
		log.Printf("home=%s peers=%d fwd=%d ret=%d drops=%d", homeStr, len(r.peersByID), r.pktFwd, r.pktRet, r.drops)
		r.mu.Unlock()
	}
}

func parseINI(filename string, section string) map[string]string {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatalf("Failed to open config %s: %v", filename, err)
	}
	defer file.Close()

	config := make(map[string]string)
	scanner := bufio.NewScanner(file)
	currentSection := ""

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if len(line) == 0 || strings.HasPrefix(line, "#") || strings.HasPrefix(line, ";") {
			continue
		}
		if strings.HasPrefix(line, "[") && strings.HasSuffix(line, "]") {
			currentSection = line[1 : len(line)-1]
			continue
		}
		if currentSection == section {
			parts := strings.SplitN(line, "=", 2)
			if len(parts) == 2 {
				key := strings.TrimSpace(parts[0])
				valPart := parts[1]
				if idx := strings.IndexAny(valPart, "#;"); idx != -1 {
					valPart = valPart[:idx]
				}
				config[key] = strings.TrimSpace(valPart)
			}
		}
	}
	return config
}

func main() {
	cfgPath := flag.String("config", "config.ini", "Path to config file")
	flag.Parse()

	log.SetFlags(log.LstdFlags | log.Lmicroseconds)

	config := parseINI(*cfgPath, "relay")
	
	bind := "0.0.0.0"
	if b, ok := config["bind_addr"]; ok && b != "" {
		bind = b
	}
	port := 51820
	if pStr, ok := config["wg_port"]; ok {
		if p, err := strconv.Atoi(pStr); err == nil {
			port = p
		}
	}

	relay := NewRelay(bind, port)
	relay.run()
}

File: vps_relay.py
#!/usr/bin/env python3
"""
vps_relay.py — Relay server (runs on AWS VPS)
==============================================
Listens on two UDP sockets:
  ctrl_port (51821) — home server registers / sends keepalives here
  wg_port   (51820) — external WireGuard peers connect here

Run:
  python vps_relay.py [config.ini]

Press Ctrl+C to stop.
"""

import asyncio
import logging
import socket
import struct
import sys
import time
from dataclasses import dataclass, field
from typing import Dict, Optional, Tuple

# ── Logging ──────────────────────────────────────────────────────────────────

logging.basicConfig(
    level=logging.DEBUG,
    format="%(asctime)s  %(levelname)-8s  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger("vps")

# ── Wire protocol ─────────────────────────────────────────────────────────────
#
#  All packets start with a 4-byte magic number so we can tell our control
#  messages apart from raw WireGuard traffic.
#
#  MAGIC(4) | type(1) | payload
#
#  MSG_REGISTER   (home→vps): no payload — just announces the NAT endpoint
#  MSG_KEEPALIVE  (home→vps): no payload — keeps NAT mapping alive
#  MSG_ACK        (vps→home): no payload — confirms receipt
#  MSG_DATA_FWD   (vps→home): peer_id(4) | src_ip(4) | src_port(2) | wg_bytes
#  MSG_DATA_REPLY (home→vps): peer_id(4) | wg_bytes

MAGIC = b"\xde\xad\xbe\xef"
MSG_REGISTER = 0x01
MSG_KEEPALIVE = 0x02
MSG_ACK = 0x03
MSG_DATA_FWD = 0x10
MSG_DATA_REPLY = 0x11

HOME_TIMEOUT = 90  # s — home considered gone after this
PEER_TIMEOUT = 120  # s — drop idle external-peer mapping after this


# ── Per-external-peer state ───────────────────────────────────────────────────


@dataclass
class Peer:
    peer_id: int
    addr: Tuple[str, int]
    last_seen: float = field(default_factory=time.monotonic)


# ── Relay ─────────────────────────────────────────────────────────────────────


class VPSRelay:
    def __init__(self, cfg: dict):
        self.ctrl_port = cfg["ctrl_port"]
        self.wg_port = cfg["wg_port"]
        self.bind_addr = cfg["bind_addr"]

        self.home_ep: Optional[Tuple[str, int]] = None
        self.home_last_seen: float = 0.0

        self._peers_by_id: Dict[int, Peer] = {}
        self._peers_by_addr: Dict[Tuple[str, int], Peer] = {}
        self._next_id = 1

        self.ctrl_tx = None  # sends to home server
        self.wg_tx = None  # sends to external WG peers

    # ── Helpers ───────────────────────────────────────────────────────────────

    def _home_alive(self) -> bool:
        return (
            self.home_ep is not None
            and time.monotonic() - self.home_last_seen < HOME_TIMEOUT
        )

    def _ack(self) -> bytes:
        return MAGIC + bytes([MSG_ACK])

    def _fwd_packet(self, peer: Peer, wg_bytes: bytes) -> bytes:
        ip_b = socket.inet_aton(peer.addr[0])
        hdr = struct.pack("!I4sH", peer.peer_id, ip_b, peer.addr[1])
        return MAGIC + bytes([MSG_DATA_FWD]) + hdr + wg_bytes

    # ── Control socket handler ────────────────────────────────────────────────

    class _Ctrl(asyncio.DatagramProtocol):
        def __init__(self, r):
            self.r = r

        def connection_made(self, tx):
            self.r.ctrl_tx = tx
            log.info(f"Control socket  :{self.r.ctrl_port}")

        def datagram_received(self, data, addr):
            self.r._on_ctrl(data, addr)

        def error_received(self, exc):
            log.error(f"ctrl error: {exc}")

    def _on_ctrl(self, data: bytes, addr: Tuple[str, int]):
        if len(data) < 5 or data[:4] != MAGIC:
            return
        t = data[4]

        if t == MSG_REGISTER:
            changed = self.home_ep != addr
            self.home_ep = addr
            self.home_last_seen = time.monotonic()
            if changed:
                log.info(f"Home server registered  {addr}")
            else:
                log.debug(f"Home re-registered  {addr}")
            self.ctrl_tx.sendto(self._ack(), addr)

        elif t == MSG_KEEPALIVE:
            if addr == self.home_ep:
                self.home_last_seen = time.monotonic()
                self.ctrl_tx.sendto(self._ack(), addr)
                log.debug(f"keepalive from {addr}")
            else:
                log.warning(f"keepalive from unknown {addr}  (home={self.home_ep})")

        elif t == MSG_DATA_REPLY:
            self._on_home_reply(data[5:], addr)

    def _on_home_reply(self, payload: bytes, addr: Tuple[str, int]):
        if addr != self.home_ep:
            log.warning(f"DATA_REPLY from unexpected {addr}")
            return
        if len(payload) < 5:
            return
        peer_id = struct.unpack("!I", payload[:4])[0]
        wg_bytes = payload[4:]
        peer = self._peers_by_id.get(peer_id)
        if not peer:
            log.warning(f"DATA_REPLY for unknown peer_id={peer_id}")
            return
        peer.last_seen = time.monotonic()
        self.wg_tx.sendto(wg_bytes, peer.addr)
        log.debug(f"→ {peer.addr}  {len(wg_bytes)} B  (id={peer_id})")

    # ── WireGuard socket handler ───────────────────────────────────────────────

    class _WG(asyncio.DatagramProtocol):
        def __init__(self, r):
            self.r = r

        def connection_made(self, tx):
            self.r.wg_tx = tx
            log.info(f"WireGuard socket  :{self.r.wg_port}")

        def datagram_received(self, data, addr):
            self.r._on_wg(data, addr)

        def error_received(self, exc):
            log.error(f"wg error: {exc}")

    def _on_wg(self, data: bytes, addr: Tuple[str, int]):
        if not self._home_alive():
            log.warning(f"WG packet from {addr} dropped — home not registered")
            return
        peer = self._peers_by_addr.get(addr)
        if peer is None:
            peer = Peer(peer_id=self._next_id, addr=addr)
            self._next_id += 1
            self._peers_by_id[peer.peer_id] = peer
            self._peers_by_addr[addr] = peer
            log.info(f"New WG peer  {addr}  id={peer.peer_id}")
        else:
            peer.last_seen = time.monotonic()
        pkt = self._fwd_packet(peer, data)
        self.ctrl_tx.sendto(pkt, self.home_ep)
        log.debug(f"← {addr}  {len(data)} B  → home  (id={peer.peer_id})")

    # ── Background tasks ──────────────────────────────────────────────────────

    async def _cleanup(self):
        while True:
            await asyncio.sleep(30)
            now = time.monotonic()
            expired = [
                p
                for p in list(self._peers_by_id.values())
                if now - p.last_seen > PEER_TIMEOUT
            ]
            for p in expired:
                self._peers_by_id.pop(p.peer_id, None)
                self._peers_by_addr.pop(p.addr, None)
                log.info(f"Dropped idle peer  {p.addr}  id={p.peer_id}")

    async def _status(self):
        while True:
            await asyncio.sleep(30)
            home = f"✓ {self.home_ep}" if self._home_alive() else "✗ (none)"
            log.info(f"home={home}  peers={len(self._peers_by_id)}")

    # ── Run ───────────────────────────────────────────────────────────────────

    async def run(self):
        loop = asyncio.get_running_loop()
        log.info(f"VPS relay  ctrl=:{self.ctrl_port}  wg=:{self.wg_port}")

        ctrl_tx, _ = await loop.create_datagram_endpoint(
            lambda: self._Ctrl(self),
            local_addr=(self.bind_addr, self.ctrl_port),
        )
        wg_tx, _ = await loop.create_datagram_endpoint(
            lambda: self._WG(self),
            local_addr=(self.bind_addr, self.wg_port),
        )

        try:
            await asyncio.gather(self._cleanup(), self._status())
        finally:
            ctrl_tx.close()
            wg_tx.close()


# ── Config ────────────────────────────────────────────────────────────────────


def load_config(path: str = "config.ini") -> dict:
    import configparser

    cfg = configparser.ConfigParser()
    cfg.read(path)
    s = cfg["relay"] if "relay" in cfg else {}
    return {
        "ctrl_port": int(s.get("ctrl_port", "51821")),
        "wg_port": int(s.get("wg_port", "51820")),
        "bind_addr": s.get("bind_addr", "0.0.0.0"),
    }


# ── Main ──────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    cfg_path = sys.argv[1] if len(sys.argv) > 1 else "config.ini"
    try:
        asyncio.run(VPSRelay(load_config(cfg_path)).run())
    except KeyboardInterrupt:
        log.info("Stopped.")
